test/test_torch.py

```diff
@@ -978,6 +979,100 @@ def _test_norm(self, device):
     def test_norm(self):
         self._test_norm(self, device='cpu')
 
+    @staticmethod
+    def _test_nuclear_norm_axes(self, device='cpu'):
+        def check_single_nuclear_norm(x, axes):
+            if x.is_cuda and randrange(100) < 95:
+                return  # too many cpu <==> gpu copies
+
+            a = np.array(x.cpu(), copy=False)
+            expected = np.linalg.norm(a, "nuc", axis=axes)
+
+            ans = torch.norm(x, "nuc", dim=axes)
+            self.assertTrue(ans.is_contiguous())
+            self.assertEqual(ans.shape, expected.shape)
+            self.assertTrue(np.allclose(ans.cpu(), expected, rtol=1e-04, atol=1e-04))
+
+            out = torch.zeros(expected.shape, dtype=x.dtype, device=x.device)
+            ans = torch.norm(x, "nuc", dim=axes, out=out)
+            self.assertIs(ans, out)
+            self.assertTrue(ans.is_contiguous())
+            self.assertEqual(ans.shape, expected.shape)
+            self.assertTrue(np.allclose(ans.cpu(), expected, rtol=1e-04, atol=1e-04))
+
+        for n in range(1, 5):
+            for m in range(1, 5):
```

Comment:
I think these ranges create too many test cases which is slowing down the tests. In my opinion: three kinds of test cases should suffice:
- square matrices
- tall matrices
- fat matrices

Square matrices could be of dimensions (3, 3), fat matrices could be of dimensions (3, 5) and tall matrices could be of dimensions (5, 3). In addition to these, we could have at most 2 batch dimensions (5, *, *) and (7, 5, *, *). This comes up to 9 test cases. In you case there are (2 + 4) * (4 * 4) = 96 test cases.

Reply:
With a release build, probably including `pytest` test collection:

***CPU***

```
$ /home/stefan/rel2/bin/python3 -m pytest test_torch.py -k test_sum_dim -v
...
1 passed, 494 deselected in 19.41 seconds
```

```
$ /home/stefan/rel2/bin/python3 -m pytest test_torch.py -k test_nuclear_norm -v
...
2 passed, 493 deselected in 2.05 seconds
```


***CUDA***

```
$ /home/stefan/rel2/bin/python3 -m pytest test_cuda.py -k test_det_logdet_slogdet -v
...
1 passed, 169 deselected in 6.37 seconds
```

95% random skip (with 98% random skip it is 4.74s):

```
/home/stefan/rel2/bin/python3 -m pytest test_cuda.py -k nuclear_norm -v
...
2 passed, 168 deselected in 7.07 seconds 
```

So CPU is still relatively fast.  On CUDA it definitely is one of the slower tests, but not completely an outlier. The benefit of brute force is that it catches issues like:

 https://github.com/pytorch/pytorch/issues/20452



Are random skips not allowed in the tests?  In the CPython test suite they are allowed, but practices differ of course.

Reply:
I don’t think random skips are allowed. However there is a `@slowtest` decorator available, introduced in https://github.com/pytorch/pytorch/pull/18231

Reply:
@ezyang do you have any suggestions?

Reply:
With  58c000c the CPU tests take 0.7s and the CUDA tests (with 95% skips) 2.4s.